### PR TITLE
Todo Form - #1 View

### PR DESCRIPTION
**Pull request by matej494**

### CODE REVIEW

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        setupViews()
+        setupObservers()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: Removing the observer was something that was required in prior versions of iOS, it's no longer necessary.

From the [Docs](https://developer.apple.com/documentation/foundation/notificationcenter/1407263-removeobserver):

> If your app targets iOS 9.0 and later or macOS 10.11 and later, you don't need to unregister
an observer in its dealloc method. Otherwise, you should call this method or removeObserver(_:) 
before observer or any object specified in addObserver(forName:object:queue:using:) or addObserver(_:selector:name:object:) is deallocated.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            textView.text = nil
+            textView.textColor = UIColor.black
+        }
+    }
+    
+    func textViewDidEndEditing(_ textView: UITextView) {
+        if textView.text.isEmpty { textViewIsEmpty() }
+    }
+    
+    func textViewIsEmpty() {
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: I don't feel the name of this method's right, as it doesn't imply an action that's being invoked. Based on what this method does, what do you think about naming this method `setPlaceholder(active: Bool)`, where, based on the `active` argument, you can `activate` or `deactivate` the `placeholder`?

This way, the method's even more flexible and usable, and usable in the `placeholder` property's `didSet`.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                                               name: NSNotification.Name.UIKeyboardWillHide,
+                                               object: nil)
+    }
+
+    @objc func keyboardWillShow(notification: NSNotification) {
+        guard let userInfo = notification.userInfo,
+            let endFrame = (userInfo[UIKeyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue
+            else { return }
+        let keyboardHeight = endFrame.size.height
+        keyboardSizedView.snp.updateConstraints {
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: We could do a one-liner here.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            $0.height.equalTo(keyboardHeight)
+        }
+        layoutIfNeeded()
+    }
+    
+    @objc func keyboardWillHide(notification: NSNotification) {
+        keyboardSizedView.snp.updateConstraints {
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: We could do a one-liner here.

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    // NOTE: id is set to -1 so we can detect if we're creating new or changing existing todo (to detect wich initializer is used)
+    private var todo = Todo(id: -1, title: "", completed: false, userId: 1)
+    private let todoFormView = TodoFormView.autolayoutView()
+    private lazy var completedButton: UIButton = {
+        let button = UIButton()
+        let image = todo.completed ? #imageLiteral(resourceName: "checkbox-selected") : #imageLiteral(resourceName: "checkbox-unselected")
+        button.setImage(image, for: .normal)
+        button.addTarget(self, action: #selector(completedButtonTapped), for: .touchDown)
+        button.snp.makeConstraints {
+            $0.width.height.equalTo(30)
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: A one-liner maybe?

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+//
+//  Created by Matej Korman on 31/07/2018.
+//  Copyright © 2018 Matej Korman. All rights reserved.
+//
+
+import SnapKit
+
+class TodoFormViewController: UIViewController {
+    // NOTE: id is set to -1 so we can detect if we're creating new or changing existing todo (to detect wich initializer is used)
+    private var todo = Todo(id: -1, title: "", completed: false, userId: 1)
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: I think that a more straightforward and understandable approach would be to initialise a new `Todo` instance via the `initialiser` that accepts an `optional` `Todo`, and keeping a `let isEditing: Bool` stored property (also initialised via the initialiser) that indicates whether or not the form is in `editing mode`.

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    private let todoFormView = TodoFormView.autolayoutView()
+    private lazy var completedButton: UIButton = {
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: We could set-up the button as so: 
```
private lazy var completedButton: UIButton = {
        let button = UIButton()
        button.setImage(#imageLiteral(resourceName: "checkbox-unselected"), for: .normal)
        button.setImage(#imageLiteral(resourceName: "checkbox-selected"), for: .selected)
        button.isSelected = todo.completed
        button.addTarget(self, action: #selector(completedButtonTapped), for: .touchDown)
        button.snp.makeConstraints { $0.width.height.equalTo(30) }
        return button
}()
```

The interesting parts are the following lines:
```
button.setImage(#imageLiteral(resourceName: "checkbox-unselected"), for: .normal)
button.setImage(#imageLiteral(resourceName: "checkbox-selected"), for: .selected)
```

You can set an properties based on the `UIButton`'s states, like we did above. So, for the `normal` state we want to use the `unselected image`, but for the `selected` state, we want to use the `selected image`. This way, we can just mutate the `isSelected` property and get the button to update automatically.

**Note:**
This works for other properties on the `UIButton`, such as the `title`, `titleColor` and so on.

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        let image = todo.completed ? #imageLiteral(resourceName: "checkbox-selected") : #imageLiteral(resourceName: "checkbox-unselected")
+        button.setImage(image, for: .normal)
+        button.addTarget(self, action: #selector(completedButtonTapped), for: .touchDown)
+        button.snp.makeConstraints {
+            $0.width.height.equalTo(30)
+        }
+        return button
+    }()
+    
+    init() {
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: Do we need the empty initialiser? 

We could just use 
```
init(todo: Todo) {
      self.todo = todo
      super.init(nibName: nil, bundle: nil)      
      setupView()
}
```

and modify it's argument list to `todo: Todo? = nil`.

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+private extension TodoFormViewController {
+    @objc func completedButtonTapped() {
+        todo.completed = !todo.completed
+        let image = todo.completed ? #imageLiteral(resourceName: "checkbox-selected") : #imageLiteral(resourceName: "checkbox-unselected")
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: With the previously mentioned code (with the different button images depending on the button's state), we could just use `completedButton.isSelected = todo.completed` here.

File: Marsiple/App/Todos/TodoTableViewCell.swift [hunk 10]
```diff
@@ -36,29 +41,61 @@ extension TodoTableViewCell {
...
+}
+
 private extension TodoTableViewCell {
     func setupViews() {
+        setupContainerViews()
         setupTodoLabel()
         setupCheckBoxImageView()
     }
     
+    func setupContainerViews() {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodoTableViewCell.swift]: Each view should get it's own `setup` method, regardless of how small.

File: Marsiple/App/Todos/TodoTableViewCell.swift [hunk 11]
```diff
@@ -36,29 +41,61 @@ extension TodoTableViewCell {
...
         checkBoxImageView.setContentHuggingPriority(UILayoutPriority.required, for: .horizontal)
         checkBoxImageView.snp.makeConstraints {
-            $0.leading.equalTo(todoLabel.snp.trailing).inset(-10)
-            $0.top.trailing.bottom.equalToSuperview().inset(10)
+            $0.edges.equalToSuperview().inset(10)
             $0.height.width.equalTo(20)
         }
     }
+    
+    func setupGestureRecognizers() {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodoTableViewCell.swift]: I don't think we should be doing this, as it's unnecessarily complex and, in essence, unnecessary. 

We can use `didSelectRow...` for opening the `todo form`, while we can create a callback from the `cell` to the `view controller` for the `completed state button`, using a `closure`.

File: Marsiple/App/Todos/TodoTableViewCellDelegate.swift [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+//  TodoTableViewCellDelegate.swift
+//  Marsiple
+//
+//  Created by Matej Korman on 01/08/2018.
+//  Copyright © 2018 Matej Korman. All rights reserved.
+//
+
+import UIKit
+
+protocol TodoTableViewCellDelegate: class {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodoTableViewCellDelegate.swift]: Remove completely.

File: Marsiple/App/Todos/TodosIndex.swift [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+//  TodosIndex.swift
+//  Marsiple
+//
+//  Created by Matej Korman on 02/08/2018.
+//  Copyright © 2018 Matej Korman. All rights reserved.
+//
+
+import Foundation
+
+enum TodosIndex: Int {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosIndex.swift:11]: Useful, good idea 👍

File: Marsiple/App/Todos/TodosViewController.swift [hunk 14]
```diff
@@ -21,6 +21,10 @@ class TodosViewController: UIViewController {
     required init?(coder aDecoder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
+
+    override func viewWillAppear(_ animated: Bool) {
+        reloadData()
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Rather than `forcing a reload` on every appearance, I would like a more precise implementation where we don't need to do a `Wild West reload` to get correct state. 

What I mean is, I would like you to implement concise callbacks for each data modifying event and sync it individiually with the table view. This requires more concise coding, but is worth it, as you fully and precisely have control over what's happening.

File: Marsiple/App/Todos/TodoTableViewCell.swift [hunk 15]
```diff
@@ -9,12 +9,17 @@
 import SnapKit
 
 class TodoTableViewCell: UITableViewCell {
-    private let checkBoxImageView = UIImageView.autolayoutView()
+    weak var delegate: TodoTableViewCellDelegate?
+    var indexPath = IndexPath()
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodoTableViewCell.swift]: This should be `an optional`, as it's not correctly available at all times.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 16]
```diff
@@ -62,13 +58,37 @@ extension TodosViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
         if editingStyle == UITableViewCellEditingStyle.delete {
             todos[indexPath.section].remove(at: indexPath.row)
-            tableView.reloadData()
+            tableView.performBatchUpdates({ tableView.deleteRows(at: [indexPath], with: .automatic) }, completion: nil)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Apply code formatting comments from the previous PRs.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 17]
```diff
@@ -62,13 +58,37 @@ extension TodosViewController: UITableViewDelegate {
...
+        navigationController?.pushViewController(TodoFormViewController(todo: todos[indexPath.section][indexPath.row]), animated: true)
+    }
+
+    func tableViewCell(didSelectCheckBoxAt indexPath: IndexPath) {
+        // TODO: Save changes on the server
+        let finishIndexPath = IndexPath(row: indexPath.section * todos[1 - indexPath.section].count, section: 1 - indexPath.section)
+        let removedTodo = todos[indexPath.section].remove(at: indexPath.row)
+        removedTodo.completed = !removedTodo.completed
+        todos[finishIndexPath.section].insert(removedTodo, at: finishIndexPath.row)
+        todosView.tableView.performBatchUpdates({
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Apply code formatting comments from the previous PRs and move to `didSelectRow..`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 18]
```diff
@@ -80,9 +100,32 @@ private extension TodosViewController {
...
+        var uncompleted = [Todo]()
+        var completed = [Todo]()
+        for i in 0...9 {
+            uncompleted.append(Todo(id: i, title: "Todo \(0)\(i)", completed: false, userId: -1))
+            completed.append(Todo(id: i, title: "Todo \(1)\(i)", completed: true, userId: -1))
+        }
         return [uncompleted, completed]
     }
+    
+    func reloadData() {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: This is the `Wild West` approach that I mentioned earlier. 

I call it like that because, what you do here is that you attempt to restore to a stable state using a set of rules (filtering todos, to get the correct state). I ask you this, why should a `completed` todo `EVER END UP` in the `active` section? That's invalid logic right at the start.

The `todos` should be modified by a concise and isolated set of operations, that, once completed, are fully finished - both on the `UI` and `Business logic` side.

File: Marsiple/Common/Models/Todo.swift [hunk 19]
```diff
@@ -8,7 +8,7 @@
 
 import Foundation
 
-struct Todo {
+class Todo {
```
Comment (dkulundzic) [inline Marsiple/Common/Models/Todo.swift]: This change does have consequences, why the change?

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Looking good, but we could make some changes to make it look great 🙂

File: Marsiple/App/Todos/TodosViewController.swift (hunk 18 above)
Comment (matej494) [reply Marsiple/App/Todos/TodosViewController.swift]: There is a reason why a completed `Todo` should end up in uncompleted `Todo`s. For instance what if user accidentally taps on the wrong `Todo`? I believe that he would be frustrated to figure out there is no way to return that `Todo` to uncompleted section. Another example would be if someone has daily chores, and at the end of the day he/she wants to reset it for the next day instead of creating new ones.

File: Marsiple/Common/Models/Todo.swift (hunk 19 above)
Comment (matej494) [reply Marsiple/Common/Models/Todo.swift]: I've done that so changes made in `TodoFormViewController` reflect on the object that `viewController` was `initialized` with.

Comment (dkulundzic) [reply Marsiple/Common/Models/Todo.swift]: This is risky behaviour. Isn't it more safe to work with two completely isolated instances and then just ovewrite the legacy one with the new one, rather than modifying the same instance from two different places?

I would argue that, even though it might seem like a good idea to use a reference type here, it isn't - a `immutable` struct's almost always a better idea.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 20]
```diff
@@ -94,13 +85,11 @@ private extension TodoFormView {
     }
     
     func setupTextView() {
-        textViewIsEmpty()
+        placeholder = ""
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: Why are we setting the `placeholder` value again? It will always have an initial value per `var placeholder = LocalizationKey.TodoForm.titlePlaceholder.localized()`.

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 21]
```diff
@@ -9,27 +9,29 @@
...
-    init(todo: Todo) {
-        self.todo = todo
+    init(todo: Todo? = nil) {
+        if let todo = todo {
+            self.todo = todo
+            isEditingTodo = true
+        } else {
+            self.todo = Todo(id: -1, title: "", completed: false, userId: -1)
+            isEditingTodo = false
+        }
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: We can shorten this quite a bit:
```
self.todo = todo ?? Todo(id: -1, title: "", completed: false, userId: -1)
self.isEditingTodo = todo != nil
```

File: Marsiple/App/Todos/TodosViewController.swift [hunk 22]
```diff
@@ -40,13 +36,19 @@ extension TodosViewController: UITableViewDataSource {
...
+        cell.didSelectCheckBox = didSelectCheckBox(atIndexPath: indexPath)
         return cell
     }
 }
 
 extension TodosViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let todoFormViewController = TodoFormViewController(todo: todos[indexPath.section][indexPath.row])
+        todoFormViewController.todoCompletedIsEdited = didSelectCheckBox(atIndexPath: indexPath)
+        todoFormViewController.todoTextIsEdited = didChangeText(atIndexPath: indexPath)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Lets create a closure here with `[weak self]` and call the `didSelectCheckBox` method directly, without it returning a `closure`, but acting directly.

For both the `todoCompletedIsEdited` and `todoTextIsEdited`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 23]
```diff
@@ -58,28 +60,12 @@ extension TodosViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
         if editingStyle == UITableViewCellEditingStyle.delete {
             todos[indexPath.section].remove(at: indexPath.row)
-            tableView.performBatchUpdates({ tableView.deleteRows(at: [indexPath], with: .automatic) }, completion: nil)
+            tableView.performBatchUpdates({ tableView.deleteRows(at: [indexPath], with: .automatic) }
+                , completion: nil)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Hmm, this doesn't look like the format we talked about:
```
tableView.performBatchUpdates({ 
          tableView.deleteRows(at: [indexPath], with: .automatic)
}, completion: nil)
```

File: Marsiple/App/Todos/TodosViewController.swift [hunk 24]
```diff
@@ -115,17 +101,52 @@ private extension TodosViewController {
         return [uncompleted, completed]
     }
     
-    func reloadData() {
-        let newlyCompleted = todos[TodosIndex.uncompleted.rawValue].filter { $0.completed }
-        if !newlyCompleted.isEmpty {
-            todos[TodosIndex.uncompleted.rawValue] = todos[TodosIndex.uncompleted.rawValue].filter { $0.completed == false }
-            todos[TodosIndex.completed.rawValue] += newlyCompleted
+    func didSelectCheckBox(atIndexPath indexPath: IndexPath) -> (() -> Void)? {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: This is unnecessary - lets just have a method that contains the code that you return here as a `closure`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 25]
```diff
@@ -115,17 +101,52 @@ private extension TodosViewController {
...
-        if !newlyUncompleted.isEmpty {
-            todos[TodosIndex.completed.rawValue] = todos[TodosIndex.completed.rawValue].filter { $0.completed }
-            todos[TodosIndex.uncompleted.rawValue] += newlyUncompleted
+        for i in indexPath.row...maxIndexOfTodos {
+            indexPaths.append(IndexPath(row: i, section: indexPath.section))
+        }
+        return indexPaths
+    }
+    
+    func didChangeText(atIndexPath indexPath: IndexPath) -> (() -> Void)? {
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Same as `didSelectCheckBox`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 26]
```diff
@@ -115,17 +101,52 @@ private extension TodosViewController {
...
-            todos[TodosIndex.uncompleted.rawValue] = todos[TodosIndex.uncompleted.rawValue].filter { $0.completed == false }
-            todos[TodosIndex.completed.rawValue] += newlyCompleted
+    func didSelectCheckBox(atIndexPath indexPath: IndexPath) -> (() -> Void)? {
+        let didSelectCheckBox = { [weak self] in
+            // TODO: Save changes on the server
+            guard let strongSelf = self
+                else { return }
+            
+            let finishIndexPath = IndexPath(row: indexPath.section * strongSelf.todos[1 - indexPath.section].count, section: 1 - indexPath.section)
+            var reloadTableView: (Bool) -> Void
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Sorry, I'm not sure I follow - why is the `reloadTableView` closure different depending on the `section`? In both cases, we should just `move` and `reload`, only the `input index paths` change.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 27]
```diff
@@ -32,27 +34,20 @@ class TodoFormView: UIView {
...
     
     func textViewDidEndEditing(_ textView: UITextView) {
-        if textView.text.isEmpty { textViewIsEmpty() }
-    }
-    
-    func textViewIsEmpty() {
-        textView.text = LocalizationKey.TodoForm.titlePlaceholder.localized()
-        textView.textColor = UIColor.lightGray
+        if textView.text.isEmpty {
+            placeholder = ""
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: Hmm, why are we setting the `placeholder` to an empty `string` when the user stops editing?

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 28]
```diff
@@ -9,27 +9,29 @@
...
+    private var todo: Todo
+    private let isEditingTodo: Bool
     private let todoFormView = TodoFormView.autolayoutView()
     private lazy var completedButton: UIButton = {
         let button = UIButton()
-        let image = todo.completed ? #imageLiteral(resourceName: "checkbox-selected") : #imageLiteral(resourceName: "checkbox-unselected")
-        button.setImage(image, for: .normal)
+        button.setImage(#imageLiteral(resourceName: "checkbox-unselected"), for: .normal)
+        button.setImage(#imageLiteral(resourceName: "checkbox-selected"), for: .selected)
+        button.isSelected = todo.completed
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: Nice 👍

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 29]
```diff
@@ -9,27 +9,29 @@
 import SnapKit
 
 class TodoFormViewController: UIViewController {
-    // NOTE: id is set to -1 so we can detect if we're creating new or changing existing todo (to detect wich initializer is used)
-    private var todo = Todo(id: -1, title: "", completed: false, userId: 1)
+    var todoTextIsEdited: (() -> Void)?
+    var todoCompletedIsEdited: (() -> Void)?
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: I think we should have `one callback closure` per high-level operation - `Creation`, `Edit`.

Something along these lines: 
```
var todoCreationHandler: ((Todo) -> Void)?
var todoEditHandler: ((Todo) -> Void)?
```
or
```
var todoHandler: ((Todo, Bool) -> Void)?  
```

where `(Todo, Bool)` represents the `Todo` that modified or created and the `Bool` represents `isEditing`.

File: Marsiple/App/Todo Form/TodoFormViewController.swift [hunk 30]
```diff
@@ -41,14 +43,22 @@ class TodoFormViewController: UIViewController {
...
         // TODO: Save changes on the server
-        todo.title = todoFormView.text
+        if isEditingTodo {
+            if todo.completed != completedButton.isSelected {
+                todo.title = todoFormView.text
+                todoCompletedIsEdited?()
+            } else if todo.title != todoFormView.text {
+                todo.title = todoFormView.text
+                todoTextIsEdited?()
+            }
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormViewController.swift]: To be modified after the discussion on the topic of callback handlers.

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Looks good, left remarks and suggestions.

File: Marsiple/App/Todo Form/TodoFormView.swift (hunk 20 above)
Comment (matej494) [reply Marsiple/App/Todo Form/TodoFormView.swift]: We're setting it so `didSet` method executes.

Comment (dkulundzic) [reply Marsiple/App/Todo Form/TodoFormView.swift]: Any initial initialisation should be done - initially, within `setupViews`. Regardless, wouldn't `placeholder = ""` terminate the value that set `LocalizationKey.TodoForm.titlePlaceholder.localized()`?

Comment (matej494) [reply Marsiple/App/Todo Form/TodoFormView.swift]: Yes it would, but that value isn't important. I could've set it to `""`. There was ```                textView.text = LocalizationKey.TodoForm.titlePlaceholder.localized()
                textView.textColor = .lightGray``` there, but i figured it would be shorter if I would just use it like this `placeholder = ""`.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 31]
```diff
@@ -38,16 +31,13 @@ class TodoFormView: UIView {
 
 extension TodoFormView: UITextViewDelegate {
     func textViewDidBeginEditing(_ textView: UITextView) {
-        if textView.textColor == .lightGray {
-            textView.text = nil
-            textView.textColor = .black
+        if textView.text == placeholder {
+            isPlaceholderActive(false)
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: What do you think about calling `setPlaceholderActive(false, newValue: textView.text)` here? When the user starts editing, `in all circumstances`, the `text view` should have its `placeholder` disabled.

File: Marsiple/App/Todo Form/TodoFormView.swift [hunk 32]
```diff
@@ -100,4 +90,14 @@ private extension TodoFormView {
             $0.height.equalTo(0)
         }
     }
+    
+    func isPlaceholderActive(_ isActive: Bool, newValue: String? = nil) {
```
Comment (dkulundzic) [inline Marsiple/App/Todo Form/TodoFormView.swift]: I would rename `newValue` to `text`, as it represents the text view's `text`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 33]
```diff
@@ -101,52 +105,40 @@ private extension TodosViewController {
...
+        todosView.tableView.performBatchUpdates({
+            todosView.tableView.moveRow(at: startIndexPath, to: finishIndexPath)
+        }, completion: { [weak self] _ in
+            self?.todosView.tableView.reloadData()
+        })
+    }
+
+    func todoIsEdited(atIndexPath indexPath: IndexPath, newValue todo: Todo) {
+        // TODO: Save changes on the server
+        todos[indexPath.section][indexPath.row].title = todo.title
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift:132]: Shouldn't we update the `Todo`'s `completed` property too?

File: Marsiple/App/Todos/TodosViewController.swift [hunk 34]
```diff
@@ -101,52 +105,40 @@ private extension TodosViewController {
...
     }
     
-    func didChangeText(atIndexPath indexPath: IndexPath) -> (() -> Void)? {
-        return { [weak self] in
-            self?.todosView.tableView.reloadRows(at: [indexPath], with: .automatic)
+    func moveRow(at startIndexPath: IndexPath, to finishIndexPath: IndexPath) {
+        todosView.tableView.performBatchUpdates({
+            todosView.tableView.moveRow(at: startIndexPath, to: finishIndexPath)
+        }, completion: { [weak self] _ in
+            self?.todosView.tableView.reloadData()
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: I would like to see the `cell` that was `moved` `reloaded` right after the `move` completes. Finally, after the `reload` completes, we need to `reload the whole table view`. 

Hint: This will be messy a bit, since you will need to nest multiple `tableView.performBatchUpdates` calls.

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Looks good! 👍 

I like how you handled the different contexts from which you can open the `TodoFormViewController` (tapping the cell or tapping `+`), setting different `todoHandlers`. 

Additionally, we lost the animated reload, once a `Todo` is completed or made active. Lets bring that back, while still keeping the `tableView`'s structural integrity consistent with a `reloadData`.

File: Marsiple/App/Todo Form/TodoFormView.swift (hunk 31 above)
Comment (matej494) [reply Marsiple/App/Todo Form/TodoFormView.swift]: @dkulundzic That wouldn't work when `placeholder` is active, because then it would just color the `placeholder` text black. I could add `if newValue == placeholder` to the method, but I don't know if that is a better solution.

File: Marsiple/App/Todos/TodosViewController.swift (hunk 33 above)
Comment (matej494) [reply Marsiple/App/Todos/TodosViewController.swift:132]: Than the following condition would never be satisfied.

File: Marsiple/App/Todos/TodosViewController.swift (hunk 34 above)
Comment (matej494) [reply Marsiple/App/Todos/TodosViewController.swift]: Why do I need to nest multiple `tableView.performBatchUpdates`? I believe that we would get the same result if `self?.todosView.tableView.reloadRows(at: [finishIndexPath], with: .automatic)` was added before `self?.todosView.tableView.reloadData()`, just like it used to be.

File: Marsiple/App/Todos/TodosViewController.swift (hunk 33 above)
Comment (dkulundzic) [reply Marsiple/App/Todos/TodosViewController.swift:132]: It would be easy to do `let shouldMove = todos[indexPath.section][indexPath.row].completed != todo.completed`, that's not the problem. 

Lets leave it as-is currently.

File: Marsiple/App/Todos/TodosViewController.swift (hunk 34 above)
Comment (dkulundzic) [reply Marsiple/App/Todos/TodosViewController.swift]: You're right, we can do that.

File: Marsiple/App/Todo Form/TodoFormView.swift (hunk 31 above)
Comment (dkulundzic) [reply Marsiple/App/Todo Form/TodoFormView.swift]: How about `isPlaceholderActive(false, newValue: textView.text == placeholder ? nil: textView.text)`?

Comment (matej494) [reply Marsiple/App/Todo Form/TodoFormView.swift]: Ok, that would work.

**Review by dkulundzic — APPROVED** [review]
Looks good! 👍 🚀

**Comment by matej494** [comment]
Successfully merged branch `todo-form/view` into branch `todo-form/main` without conflicts.
